src/Inciter/Diagnostics.C

```diff
@@ -0,0 +1,141 @@
+    // that is done by calling PDE::initialize(). If the analytical solution is
+    // a function of time, that is already incorporated in setting initial
+    // conditions. For those PDEs where the analytical solution is not
+    // available, initialize() returns the initial conditions (obviously), and
+    // thus the "error", defined between this "analytical" and numerical
+    // solution will be a measure of the "distance" between the initial
+    // condition and the current numerical solution. This is not necessarily
+    // useful, but simplies the logic because all PDEs can be treated as being
+    // able to compute an error based on some "analytical" solution, which is
+    // really the initial condition.
+    auto a = u;
+    for (const auto& eq : g_pdes)
+      eq.initialize( d.Coord(), a, d.T()+d.Dt(), d.Gid() );
+
+    // Prepare for computing diagnostics. Diagnostics are defined as some norm,
+    // .e.g., L2 norm, of a quantity, computed in mesh nodes, A, as ||A||_2 =
+    // sqrt[ sum_i(A_i)^2 V_i ], where the sum is taken over all mesh nodes and
+    // V_i is the nodal volume. We send multiple sets of quantities to the host
+    // for aggregation across the whole mesh: (1) the numerical solutions of all
+    // components of all PDEs, and their error, defined as A_i = (a_i - n_i),
+    // where a_i and n_i are the analytical and numerical solutions at node i,
+    // respectively. The final aggregated solution will end up in
+    // Transporter::diagnostics().
+
+    // Diagnostics vector (of vectors) during aggregation:
+    // 0: L2-norm of all scalar components of the numerical solution
+    // 1: L2-norm of all scalar components of the numerical-analytic solution
+    // 2: Linf-norm of all scalar components of the numerical-analytic solution
+    std::vector< std::vector< tk::real > >
+      diag( 6, std::vector< tk::real >( u.nprop(), 0.0 ) );
+
+    // Put in norms sweeping our mesh chunk
+    for (std::size_t i=0; i<u.nunk(); ++i)
+      if (m_slave.find(i) == end(m_slave)) {    // ignore non-owned nodes
+        // Compute L2 norm of the numerical solution
+        for (std::size_t c=0; c<u.nprop(); ++c)
+          diag[0][c] += u(i,c,0) * u(i,c,0) * d.Vol()[i];
+        // Compute L2 norm of the numerical-analytic solution
+        for (std::size_t c=0; c<u.nprop(); ++c)
+          diag[1][c] +=
+            (u(i,c,0)-a(i,c,0)) * (u(i,c,0)-a(i,c,0)) * d.Vol()[i];
+        // Compute Linf norm of the numerical-analytic solution
+        for (std::size_t c=0; c<u.nprop(); ++c) {
+          auto err = std::abs( u(i,c,0) - a(i,c,0) );
+          if (err > diag[2][c]) diag[2][c] = err;
+        }
+      }
+
+    // Append diagnostics vector with metadata on the current time step
+    // 3: Current iteration count (only the first entry is used)
+    // 4: Current physical time (only the first entry is used)
+    // 5: Current physical time step size (only the first entry is used)
```

Comment:
Could be an enum or similar, as part of removing the magic numbers?

Reply:
Sure: 7f06ea79

Reply:
Sure: 7f06ea79.

Reply:
Also, I have added some more meaningful comments in 7f06ea79 that clarify that at that point it is only a sum/max as the first step in computing the norms.